### DISCUSSION TITLE
Help tracking down a pandoc error: Cannot decode byte in a Quarto project

### DISCUSSION THREAD

**Discussion by friendly** [quarto-dev/quarto-cli#5787] [Q&A]
### Description

**Please help** I'm getting ready to abandon Quarto.

In [this StackOverflow post](https://stackoverflow.com/questions/76366976/quarto-pandoc-cant-find-source-of-pandoc-error-pandoc-exe-cannot-decode) I describe a problem with pandoc (v. 2.19.2) under quarto (v. 1.1.189), and RStudio (RStudio 2022.07.2+576)where I get this error,

```
pandoc.exe: Cannot decode byte '\x93': Data.Text.Internal.Encoding.decodeUtf8: Invalid UTF-8 stream
```

and have been persistently unable to find the location in my files where this might occur. As I noted in SO, I have checked all my `.qmd`, `.R`, `'.bib`, files using `tools::showNonASCIIfile()` and nothing turns up.

I also filed this as a [pandoc issue](https://github.com/jgm/pandoc/issues/8884).  I updated `pandoc` to the latest version (3.1.2)
and reduced my book `index.qmd` to the bare minimum:

```
# Preface {.unnumbered}
```

In my `_quarto.yml` I've commented out all chapters other than `index.qmd`

```
  chapters:
    - index.qmd
    # - intro.qmd
    # - getting_started.qmd
    # - multivariate_plots.qmd
    # - linear_models.qmd
```

* Are there any tools to debug or trace such a problem?
* In the quarto internals, is there any way to provide more information on such a pandoc error?

To help track this down, I've made my repo public, at https://github.com/friendly/Vis-MLM-quarto

**Comment by friendly** [comment]
I have also checked the encoding of my files in RStudio.  They seem to be UTF-8. Is there any way I can check all the files in my book folder, rather than one by one?

I also checked my Windows 10 `chcp` setting.  It gives **850**, but not sure what this means.

All of this worked up until a couple of days ago, and I don't think anything changed on my system

**Comment by cscheid** [comment]
> under quarto (v. 1.1.189)

This is a relatively old version of quarto; can you first try [upgrading to our latest stable version](https://quarto.org/docs/get-started/) and see if the problem persists?

**Reply by friendly** [reply]
Thank you for your reply.
I've just:

* upgraded quarto to the latest, 1.3.361
* upgraded pandoc from the main pandoc site to pandoc.exe 3.1.2
* wiped out my local repo and re-cloned from GitHub

I still get the same error:
```
Rendering:
[1/1] index.qmd
pandoc: Cannot decode byte '\x93': Data.Text.Internal.Encoding.decodeUtf8: Invalid UTF-8 stream
```

**Reply by cscheid** [reply]
> upgraded pandoc from the main pandoc site to pandoc.exe 3.1.2

For what's worth, you shouldn't need to upgrade your local pandoc; quarto ships its own version. What's the output you get from `quarto check`?

In addition, since your repository is publicly available, would you mind adding

````
---
keep-md: true
---
````

to your `index.qmd` file, running `quarto render` again, and then checking in `index.md` to the git repository? This way we might be able to inspect the file for issues.

**Reply by friendly** [reply]
OK, now I think we're getting somewhere, for which I'm deeply grateful.

Here's `quarto check`.  Does anything look strange?
 I'll set `keep-md: true` shortly

```
friendly@UIT-HHP-2PQL063 MINGW64 /c/R/Projects/Vis-MLM-quarto (master)
$ quarto check
A new release of Deno is available: 1.28.2 ÔåÆ 1.34.1 Run `deno upgrade` to install it.

[>] Checking versions of quarto binary dependencies...
      Pandoc version 3.1.1: OK
      Dart Sass version 1.55.0: OK
[>] Checking versions of quarto dependencies......OK
[>] Checking Quarto installation......OK
      Version: 1.3.361
      Path: C:\Program Files\Quarto\bin
      CodePage: unknown

[>] Checking basic markdown render....OK

[>] Checking Python 3 installation....(None)

      Unable to locate an installed version of Python 3.
      Install Python 3 from https://www.python.org/downloads/

[>] Checking R installation...........OK
      Version: 4.2.3
      Path: C:/R/R-4.2.3
      LibPaths:
        - C:/R/R-4.2.3/library
      knitr: 1.43
      rmarkdown: 2.21

[>] Checking Knitr engine render......OK


friendly@UIT-HHP-2PQL063 MINGW64 /c/R/Projects/Vis-MLM-quarto (master)
$
```

**Reply by friendly** [reply]
`keep-md: true` done and pushed to the repo.  

`index.md` contains only one line:

```
# Preface {.unnumbered}
```

**Reply by friendly** [reply]
Hi Carlos
Is there anything further I can test to find the source of this problem?
Is it possible that someone could download my repo and try to build the document?
I'm gratefujl for any help with this.

**Reply by dragonstyle** [reply]
I tried this locally on both a MacOS and Windows machine, both of which were able to render the file fine.

Is there anything unique or unusual about your configuration that might somehow be related? Do you have access to an alternate machine that you could try this on?

Two things occur to me:

- Something about the environment is different in some that we don't understand and that is somehow unearthing the bug (e.g. file system code page, shell / shell behavior, etc...)
- Somehow other data is being included in the rendering (for example from a pandoc `data-dir` - we attempt to lock down the `data-dir` to our own, but maybe somehow something is sneaking in through a crack we're not aware of).

Those are my only current theories - I'm sorry I don't have better guesses!

**Reply by cscheid** [reply]
The `MINGW64` line makes me suspect it's something about the localization settings of that terminal itself.

Note also that the `ÔåÆ` garbage characters are probably unicode being badly interpreted. I think that has to be the source of the problem.

**Reply by cderv** [reply]
Good catch @cscheid ! To try isolate this, @friendly is this possible to run `quarto render` in another terminal like powershell that should be available on Windows by default ? 

If you have no issue with it, then it should be something related to running Pandoc in this specific bash emulation environment.

**Reply by friendly** [reply]
Thank you all for the help. Very much appreciated, guys!

I finally found the source of this error -- a bib file outside my project that was imported and not checked. Yet, it took me over 2 days, and could have been avoided if pandoc returned filename and file location information in its error message.

I filed this as an enhancement request at pandoc:, https://github.com/jgm/pandoc/issues/8884

**Reply by cderv** [reply]
Wow good catch ! This is a bad one, and Pandoc should indeed help with the error ! Thanks for filling that up.
